merlin/models/torch/container.py

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Iterator, Optional, Union
+
+from torch import nn
+from torch._jit_internal import _copy_to_script_wrapper
+
+from merlin.models.torch.utils import torchscript_utils
+
+
+class BlockContainer(nn.Module):
+    """A container class for PyTorch `nn.Module` that allows for manipulation and traversal
+    of multiple sub-modules as if they were a list. The modules are automatically wrapped
+    in a TorchScriptWrapper for TorchScript compatibility.
+
+    Parameters
+    ----------
+    *inputs : nn.Module
+        One or more PyTorch modules to be added to the container.
+    name : Optional[str]
+        An optional name for the BlockContainer.
+    """
+
+    def __init__(self, *inputs: nn.Module, name: Optional[str] = None):
+        super().__init__()
+        self.values = nn.ModuleList()
+
+        for module in inputs:
+            self.values.append(self.wrap_module(module))
+
+        self._name: str = name
+
+    def append(self, module: nn.Module):
+        """Appends a given module to the end of the list.
+
+        Parameters
+        ----------
+        module : nn.Module
+            The PyTorch module to be appended.
+
+        Returns
+        -------
+        self
+        """
+        self.values.append(self.wrap_module(module))
+
+        return self
+
+    def prepend(self, module: nn.Module):
+        """Prepends a given module to the beginning of the list.
+
+        Parameters
+        ----------
+        module : nn.Module
+            The PyTorch module to be prepended.
+
+        Returns
+        -------
+        self
+        """
+        return self.insert(0, module)
```

Comment:
Should we wrap the `module` here with `self.wrap_module`?

Reply:
Right now the wrapping of the module happens inside the `insert` method. This might be confusing. Do you think it's cleaner if we do it in 2 places? (The 2nd time that would amount to a no-op since it's already wrapped).

Reply:
On second look, this looks fine. I think I confused `self.insert` with `self.values.insert`.